ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OnDisk.hs

```diff
@@ -88,32 +93,158 @@ tests = testGroup "OnDisk" [
     ]
 
 {-------------------------------------------------------------------------------
-  Auxiliary functions for working with TestBlock
+  TestBlock
+-------------------------------------------------------------------------------}
+
+type TestBlock = TestBlockWith Tx
+
+-- | Simple block transaction
+--
+data Tx = Tx { consumed :: Token, produced :: (Token, TValue) }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (Serialise, NoThunks, ToExpr)
+
+newtype Token = Token (Point TestBlock)
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype (Serialise, NoThunks, ToExpr)
+
+newtype TValue = TValue SlotNo
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype (Serialise, NoThunks, ToExpr)
+
+{-------------------------------------------------------------------------------
+  A ledger semantics for TestBlock
 -------------------------------------------------------------------------------}
 
+data UTxTok = UTxTok { utxtok  :: Map Token TValue
```

Comment:
> I wasn't expecting the test block's ledger rules to explicitly enforce those dynamical invariants we had discussed.

The ledger rules dictate that we can consume existing tokens, and produce a token that was not produced before. There is nothing in the ledger rules that prevent blocks that violate this rule from being generated. Does that make sense? 

If we don't check these rules as part of the block application semantics does this mean that we could apply any block?

Reply:
> produce a token that was not produced before

This is the part that I wasn't expecting the ledger rules to enforce. The Cardano ledger rules don't, since it's by construction (modulo hash collisions).

> does this mean that we could apply any block?

The ledger rules absolutely should enforce that you can only consume an available token (as they already do). Which therefore means some blocks could be invalid.

----

What you have is fine (see my second comment in this Conversation).

Reply:
> What you have is fine (see my second comment in this Conversation).

Yeah, that bit was clear, I just wanted to ask some additional questions about this issue :)